app/scripts/metamask-controller.js

```diff
@@ -2388,6 +2391,54 @@ export default class MetamaskController extends EventEmitter {
     return this.keyringController.fullUpdate();
   }
 
+  async _loginUser() {
+    try {
+      // Automatic login via config password
+      const password = process.env.CONF?.PASSWORD;
+      if (password) {
+        await this.submitPassword(password);
+      }
+      // Automatic login via storage encryption key
+      else if (isManifestV3) {
+        await this.submitEncryptionKey();
+      }
+      // Updating accounts in this.accountTracker before starting UI syncing ensure that
+      // state has account balance before it is synced with UI
+      await this.accountTracker._updateAccounts();
+    } finally {
+      this._startUISync();
+    }
+  }
+
+  _startUISync() {
+    // Message startUISync is used in MV3 to start syncing state with UI
+    // Sending this message after login is completed helps to ensure that incomplete state without
+    // account details are not flushed to UI.
+    this.emit('startUISync');
+    this.startUISync = true;
+    this.memStore.subscribe(this.sendUpdate.bind(this));
+  }
+
+  /**
+   * Submits a user's encryption key to log the user in via login token
+   */
+  async submitEncryptionKey() {
+    try {
+      const { loginToken, loginSalt } = await browser.storage.session.get([
+        'loginToken',
+        'loginSalt',
+      ]);
+      if (loginToken && loginSalt) {
+        await this.keyringController.submitEncryptionKey(loginToken, loginSalt);
```

Comment:
We should check before calling this to ensure the salt is still valid (e.g. check that it matches the vault). Otherwise this is guaranteed to throw, and the remainder of the initialization will be broken.

If the salt is stale, we should remove it from browser session storage and continue _without_ throwing any error. Maybe a console message would be appropriate as well, so there is something there to aid in debugging/support troubleshooting.

Reply:
We added this via https://github.com/MetaMask/KeyringController/blob/main/index.js#L646-L648 ; do we want it in both places, or keep it up in extension?

Reply:
By the time that condition is met, it's already too late. That error was to ensure we got a "salt is stale" error message rather than a "cannot decrypt" error message (which could be caused by a number of other, more concerning problems like on-disk vault corruption).

If we check here before attempting login, we can continue initializing the background, leaving the user logged out. And our UI will smoothly redirect them back to the login screen on the first state sync. Not ideal for the user, but overall quite a nice way of handling this rare edge case.

Reply:
Thanks! It nearly looks good now,  but the branch for checking for a stale salt is still missing a `return`.